pgp/keysource.go

```diff
@@ -25,259 +25,424 @@ import (
+//  as it becomes difficult to sugar the logger with data for e.g. individual
+//  processes.
 var log *logrus.Logger
 
 func init() {
 	log = logging.NewLogger("PGP")
 }
 
-// MasterKey is a PGP key used to securely store sops' data key by encrypting it and decrypting it
+// MasterKey is a PGP key used to securely store SOPS' data key by
+// encrypting it and decrypting it.
 type MasterKey struct {
-	Fingerprint  string
+	// Fingerprint contains the fingerprint of the PGP key used to Encrypt
+	// or Decrypt the data key with.
+	Fingerprint string
+	// EncryptedKey contains the SOPS data key encrypted with PGP.
 	EncryptedKey string
+	// CreationDate of the MasterKey, used to determine if the EncryptedKey
+	// needs rotation.
 	CreationDate time.Time
-}
 
-// EncryptedDataKey returns the encrypted data key this master key holds
-func (key *MasterKey) EncryptedDataKey() []byte {
-	return []byte(key.EncryptedKey)
+	// gnuPGHomeDir contains the absolute path to a GnuPG home directory.
+	// It can be injected by a (local) keyservice.KeyServiceServer using
+	// GnuPGHome.ApplyToMasterKey().
+	gnuPGHomeDir string
+	// disableAgent instructs the MasterKey to not use the GnuPG agent during
+	// decryption operations.
+	disableAgent bool
+	// pubRing contains the absolute path to a public keyring used by OpenPGP.
+	// When empty, defaultPubRing relative to GnuPG home is assumed.
+	pubRing string
+	// secRing contains the absolute path to a sec keyring used by OpenPGP.
+	// When empty, defaultSecRing relative to GnuPG home is assumed.
+	secRing string
 }
 
-// SetEncryptedDataKey sets the encrypted data key for this master key
-func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
-	key.EncryptedKey = string(enc)
+// NewMasterKeyFromFingerprint takes a PGP fingerprint and returns a new
+// MasterKey with that fingerprint.
+func NewMasterKeyFromFingerprint(fingerprint string) *MasterKey {
+	return &MasterKey{
+		Fingerprint:  strings.Replace(fingerprint, " ", "", -1),
+		CreationDate: time.Now().UTC(),
+	}
 }
 
-func gpgBinary() string {
-	binary := "gpg"
-	if envBinary := os.Getenv("SOPS_GPG_EXEC"); envBinary != "" {
-		binary = envBinary
+// MasterKeysFromFingerprintString takes a comma separated list of PGP
+// fingerprints and returns a slice of new MasterKeys with those fingerprints.
+func MasterKeysFromFingerprintString(fingerprint string) []*MasterKey {
+	var keys []*MasterKey
+	if fingerprint == "" {
+		return keys
 	}
-	return binary
+	for _, s := range strings.Split(fingerprint, ",") {
+		keys = append(keys, NewMasterKeyFromFingerprint(s))
+	}
+	return keys
 }
 
-func (key *MasterKey) encryptWithGPGBinary(dataKey []byte) error {
-	fingerprint := key.Fingerprint
-	if offset := len(fingerprint) - 16; offset > 0 {
-		fingerprint = fingerprint[offset:]
+// GnuPGHome is the absolute path to a GnuPG home directory.
+// A new keyring can be constructed by combining the use of NewGnuPGHome() and
+// Import() or ImportFile().
+type GnuPGHome string
```

Comment:
I think the whole framework is based explained from the perspective of one writing their own local keyservice server implementation. Maybe as a follow up, this could be added as a code example? I have most of the code laying around for it anyway, and this would put it where I would have looked (the Go pkg docs).

Reply:
That sounds perfect!